src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -82,22 +123,82 @@ def codegen_workdir(self, target):
     be the same for the same input target.
+    :param Target target: the codegen target (e.g., a java_protobuf_library).
     :return: The absolute file path.
     """
-    # TODO(gm): This method will power the isolated/global strategies for what directories to put
-    # generated code in, once that exists. This will work in a similar fashion to the jvm_compile
-    # tasks' isolated vs global strategies, generated code per-target in a way that avoids
-    # collisions.
-    return self.workdir
+    return os.path.join(self.workdir, self._codegen_workdir_suffix(target, self.codegen_strategy))
+
+  def _execute_strategy_isolated(self, targets):
+    for target in targets:
+      self.execute_codegen([target])
+
+  def _execute_strategy_global(self, targets):
+    self.execute_codegen(targets)
+
+  def _find_sources_generated_by_target(self, target):
+    if target.id in self._generated_sources_cache:
+      for source in self._generated_sources_cache[target.id]:
+        yield source
+    target_workdir = self.codegen_workdir(target)
+    if not os.path.exists(target_workdir):
+      return
+    # breadth-first search to find generated files.
+    frontier = [target_workdir]
+    while frontier:
+      path = frontier.pop(0)
+      if os.path.isdir(path):
+        frontier.extend([os.path.join(path, file_name) for file_name in os.listdir(path)])
+      else:
+        yield path
+
+  def _find_sources_generated_by_dependencies(self, target, relative=False):
+    sources = OrderedSet()
+    def add_sources(dep):
+      if dep is not target:
+        dep_sources = self._find_sources_generated_by_target(dep)
+        if relative:
+          dep_sources = [self._relative_source(dep, source) for source in dep_sources]
+        sources.update(dep_sources)
+    target.walk(add_sources)
+    return sources
+
+  def _relative_source(self, target, source):
+    return os.path.relpath(source, self.codegen_workdir(target))
+
+  def _find_sources_strictly_generated_by_target(self, target):
+    if target.id in self._generated_sources_cache:
+      return self._generated_sources_cache[target.id]
+    by_target = OrderedSet(self._find_sources_generated_by_target(target))
+    by_dependencies = self._find_sources_generated_by_dependencies(target, relative=True)
+    strict = [t for t in by_target if self._relative_source(target, t) not in by_dependencies]
+    # TODO(gm): remove debug code or actually log it in the logger.
+    print('\n[Target] {name}\n\t  {by_target}\n[Parents]\n\t  {by_parents}\n[Strict]\n\t  {strict}'
+      .format(name=target.address.spec,
+              by_target='\n\t  '.join(os.path.basename(s) for s in by_target),
+              by_parents='\n\t  '.join(os.path.basename(s) for s in by_dependencies),
+              strict='\n\t  '.join(os.path.basename(s) for s in strict)))
+    self._generated_sources_cache[target.id] = strict
+    return strict
 
   def execute(self):
+    execute_strategies = {
```

Comment:
I think this might be better handled by creating a class that is like an interface (There is an @abstractmethod decorator in  python you could use, e.g. `from abc import abstractmethod` )  and then create two concrete classes.